modules/images/image-loading-optimization/optimization.php

```diff
@@ -0,0 +1,174 @@
+<?php
+/**
+ * Optimizing for image loading optimization.
+ *
+ * @package performance-lab
+ * @since n.e.x.t
+ */
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly.
+}
+
+require_once __DIR__ . '/class-ilo-html-tag-processor.php';
+
+/**
+ * Adds template output buffer filter for optimization if eligible.
+ *
+ * @since n.e.x.t
+ * @access private
+ */
+function ilo_maybe_add_template_output_buffer_filter() {
+	if ( ! ilo_can_optimize_response() ) {
+		return;
+	}
+	add_filter( 'ilo_template_output_buffer', 'ilo_optimize_template_output_buffer' );
+}
+add_action( 'wp', 'ilo_maybe_add_template_output_buffer_filter' );
+
+/**
+ * Constructs preload links.
+ *
+ * @since n.e.x.t
+ * @access private
+ *
+ * @param array $lcp_images_by_minimum_viewport_widths LCP images keyed by minimum viewport width, amended with attributes key for the IMG attributes.
+ * @return string Markup for zero or more preload link tags.
+ */
+function ilo_construct_preload_links( array $lcp_images_by_minimum_viewport_widths ): string {
+	$preload_links = array();
+
+	$minimum_viewport_widths = array_keys( $lcp_images_by_minimum_viewport_widths );
+	for ( $i = 0, $len = count( $minimum_viewport_widths ); $i < $len; $i++ ) {
+		$lcp_element = $lcp_images_by_minimum_viewport_widths[ $minimum_viewport_widths[ $i ] ];
+		if ( false === $lcp_element || empty( $lcp_element['attributes'] ) ) {
+			// No LCP element at this breakpoint, so nothing to preload.
+			continue;
+		}
+
+		$img_attributes = $lcp_element['attributes'];
+
+		// Prevent preloading src for browsers that don't support imagesrcset on the link element.
+		if ( isset( $img_attributes['src'], $img_attributes['srcset'] ) ) {
+			unset( $img_attributes['src'] );
+		}
+
+		// Add media query if it's going to be something other than just `min-width: 0px`.
+		if ( $minimum_viewport_widths[ $i ] > 0 || isset( $minimum_viewport_widths[ $i + 1 ] ) ) {
+			$media_query = sprintf( '( min-width: %dpx )', $minimum_viewport_widths[ $i ] );
+			if ( isset( $minimum_viewport_widths[ $i + 1 ] ) ) {
+				$media_query .= sprintf( ' and ( max-width: %dpx )', $minimum_viewport_widths[ $i + 1 ] - 1 );
```

Comment:
Related to the above, it looks like this is the only reason why the `$i` is relevant. A simpler solution would be to capture the last key in the array before the loop, then go through the `foreach` loop and instead of `isset( $minimum_viewport_widths[ $i + 1 ] )` check for `$minimum_viewport_width !== $last_viewport_width`.

Reply:
Humm. We need more than just the last key. We need the next key after the current key. So if the minimum widths are (0, 480, 782), then we need to iterate 3 times and have access to these values:

1. 0, 480-1
2. 480, 782-1
3. 782, null

Reply:
I've made this easier to read in 2758c0dba01a489e0b204cfcbb57d66b4d47b5ef. Let me know if you have anything else in mind to improve this.